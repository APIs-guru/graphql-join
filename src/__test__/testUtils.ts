declare var expect;

import {
  Source,
  DocumentNode,
  GraphQLError,
  GraphQLSchema,
  GraphQLFieldResolver,
  GraphQLResolveInfo,
  ExecutionResult,

  graphql,
  parse,
  print,
  buildSchema,
  printSchema,
  formatError,
} from 'graphql';
import * as _ from 'lodash';

import { joinSchemas, ProxyContext } from '../index';
import { stubSchema } from '../utils';

expect.addSnapshotSerializer({
  print(val) {
    return typeof val === 'string' ?
      val.trim() :
      JSON.stringify(val, null, 2);
  },
  test(val) {
    return true;
  },
});

function fakeFieldResolver(schemaName: string): GraphQLFieldResolver<any,any> {
  return (
    _1, _2, _3,
    {fieldName, parentType}: GraphQLResolveInfo
  ) => {
    return `${schemaName}::${parentType.name}::${fieldName}`;
  }
}

function makeProxy(schemaName: string, schema: GraphQLSchema) {
  return (queryAST: DocumentNode) => {
    const query = print(queryAST);
    expect(query).toMatchSnapshot();
    return graphql({
      schema,
      source: query,
      fieldResolver: fakeFieldResolver(schemaName),
    });
  };
}

type TestSchema = string;
type TestSchemasMap = { [name: string]: TestSchema };
export function testJoin(testSchemas: TestSchemasMap, joinSDL: string) {
  const remoteSchemas = _.mapValues(testSchemas, (sdl, name) => {
    const schema = buildSchema(new Source(sdl, name));
    stubSchema(schema);
    return { schema };
  });

  const joinAST = parse(new Source(joinSDL, 'Join SDL'));
  const schema = joinSchemas(joinAST, remoteSchemas);

  expect(schema).toBeInstanceOf(GraphQLSchema);
  expect(printSchema(schema)).toMatchSnapshot();
  return async (
    query: string,
    results?: { [schemaName: string]: ExecutionResult }
  ) => {
    const proxyFns = _.mapValues(remoteSchemas, ({schema}, name) => {
      if (results && results[name]) {
        return (() => Promise.resolve(results[name]));
      }
      return makeProxy(name, schema)
    });

    const result = await graphql({
      schema,
      source: new Source(query, 'ClientQuery'),
      contextValue: new ProxyContext(proxyFns),
    });
    expect([
      query,
      resultToJSON(result),
    ]).toMatchSnapshot();
  };
}

function resultToJSON(result) {
  return {
    ...result,
    errors: (result.errors || []).map(formatError),
  };
}

