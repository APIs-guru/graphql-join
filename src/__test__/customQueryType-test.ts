import { testJoin } from './testUtils';

describe('custom Query type', () => {

  test('one schema', async () => {
    const execute = testJoin({
      test: 'type Query { foo: String, bar:String }',
    }, `
      type Query {
        foo: String @resolveWith(query: "foo")
      }
      query foo @send(to: "test") { foo }
    `);
    await execute('{ foo }');
  });

  test('two schemas', async () => {
    const execute = testJoin({
      test1: 'type Query { foo: String, bar: String }',
      test2: 'type Query { baz: String }',
    }, `
      type Query {
        foo: String @resolveWith(query: "foo")
        baz: String @resolveWith(query: "baz")
      }
      query foo @send(to: "test1") { foo }
      query baz @send(to: "test2") { baz }
    `);
    await execute('{ foo }');
    await execute('{ baz }');
    await execute('{ foo baz }');
  });

});
