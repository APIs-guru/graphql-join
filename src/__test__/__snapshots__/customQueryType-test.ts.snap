// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom Query type one schema 1`] = `
type Query {
  foo: String
}
`;

exports[`custom Query type one schema 2`] = `
{
  foo
}
`;

exports[`custom Query type one schema 3`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": "test::Query::foo"
    },
    "errors": []
  }
]
`;

exports[`custom Query type two schemas 1`] = `
type Query {
  foo: String
  baz: String
}
`;

exports[`custom Query type two schemas 2`] = `
{
  foo
}
`;

exports[`custom Query type two schemas 3`] = `
[
  "{ foo }",
  {
    "data": {
      "foo": "test1::Query::foo"
    },
    "errors": []
  }
]
`;

exports[`custom Query type two schemas 4`] = `
{
  baz
}
`;

exports[`custom Query type two schemas 5`] = `
[
  "{ baz }",
  {
    "data": {
      "baz": "test2::Query::baz"
    },
    "errors": []
  }
]
`;

exports[`custom Query type two schemas 6`] = `
{
  foo
}
`;

exports[`custom Query type two schemas 7`] = `
{
  baz
}
`;

exports[`custom Query type two schemas 8`] = `
[
  "{ foo baz }",
  {
    "data": {
      "foo": "test1::Query::foo",
      "baz": "test2::Query::baz"
    },
    "errors": []
  }
]
`;
